gsd/gsd.c

```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
-        handle->index_num_entries = R;
+        buf->size = R;
     }
 
-    // determine the current frame counter
-    if (handle->index_num_entries == 0)
+    return GSD_SUCCESS;
+}
+
+/** @internal
+    @brief Free the memory allocated by the index buffer or unmap the mapped memory.
+
+    @param buf Buffer to free.
+
+    @returns GSD_SUCCESS on success, GSD_* error codes on error.
+*/
+static int gsd_index_buffer_free(struct gsd_index_buffer* buf)
+{
+    if (buf == NULL || buf->data == NULL)
     {
-        handle->cur_frame = 0;
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
+
+#if GSD_USE_MMAP
+    if (buf->mapped_data)
+    {
+        int retval = munmap(buf->mapped_data, buf->mapped_len);
+
+        if (retval != 0)
+        {
+            return GSD_ERROR_IO;
+        }
     }
     else
+#endif
     {
-        handle->cur_frame = handle->index[handle->index_num_entries - 1].frame + 1;
+        free(buf->data);
     }
 
-    // at this point, all valid index entries have been written to disk
-    handle->index_written_entries = handle->index_num_entries;
+    gsd_util_zero_memory(buf, sizeof(struct gsd_index_buffer));
+    return GSD_SUCCESS;
+}
+
+/** @internal
+    @brief Add a new index entry and provide a pointer to it.
+
+    @param buf Buffer to add too.
+    @param entry [out] Pointer to set to the new entry.
 
-    if (handle->open_flags == GSD_OPEN_APPEND)
+    Double the size of the reserved space as needed to hold the new entry. Does not accept mapped
+    indices.
+
+    @returns GSD_SUCCESS on success, GSD_* error codes on error.
+*/
+inline static int gsd_index_buffer_add(struct gsd_index_buffer* buf, struct gsd_index_entry** entry)
+{
+    if (buf == NULL || buf->mapped_data || entry == NULL || buf->reserved == 0)
     {
-#if GSD_USE_MMAP
-        // in append mode, we need to tear down the temporary mapping and allocate a temporary
-        // buffer to hold indices for a single frame
-        size_t page_size = getpagesize();
-        size_t index_size = sizeof(struct gsd_index_entry) * handle->header.index_allocated_entries;
-        size_t offset = (handle->header.index_location / page_size) * page_size;
-        int retval
-            = munmap(handle->mapped_data, index_size + (handle->header.index_location - offset));
-        handle->index = NULL;
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
 
-        if (retval != 0)
+    if (buf->size == buf->reserved)
+    {
+        // grow the array
+        size_t new_reserved = buf->reserved * 2;
+        buf->data = realloc(buf->data, sizeof(struct gsd_index_entry) * new_reserved);
+        if (buf->data == NULL)
+        {
+            return GSD_ERROR_MEMORY_ALLOCATION_FAILED;
+        }
+
+        // zero the new memory
+        gsd_util_zero_memory(buf->data + buf->reserved,
+                             sizeof(struct gsd_index_entry) * (new_reserved - buf->reserved));
+        buf->reserved = new_reserved;
+    }
```

Comment:
Can you define a `gsd_buffer_resize` function or so that can be reused in other places like the function for appending to a byte buffer?

Reply:
I could have, that might have made `gsd_flush_write_buffer` cleaner to implement. At this point, the code is implemented and working so I'm not going to change it.